File: apps/api/src/App.Api/Models/ApiContracts.cs
namespace App.Api.Models;

/// <summary>
/// HTTP contract between the SPA and the Lambda API. Kept in parity with
/// packages/shared/src/api-contracts.ts since the two applications no longer
/// share a compiler -- see docs/api.md for the cross-language contract note.
/// </summary>
public sealed record HealthResponse(string Status, string Timestamp);

/// <summary>
/// Allow-listed identity projection returned by GET /api/me. Only stable
/// identifiers and presentation data are returned; the raw token and the full
/// claim set are never echoed back to the browser.
/// </summary>
public sealed record MeResponse(
    string UserId,
    string TenantId,
    IReadOnlyList<string> Scopes,
    IReadOnlyList<string> Roles,
    string? DisplayName);

public sealed record ApplicationRecord(
    string Id,
    string Name,
    string? Description,
    string OwnerId,
    string CreatedAt);

public sealed record CreateApplicationRequest(string? Name, string? Description);

public sealed record ListApplicationsResponse(IReadOnlyList<ApplicationRecord> Items);

public sealed record ApiErrorDetail(
    string Code,
    string Message,
    string CorrelationId,
    IReadOnlyDictionary<string, string>? Details);

public sealed record ApiErrorBody(ApiErrorDetail Error);

File: packages/shared/src/api-contracts.ts
/**
 * HTTP contract between the SPA and the Lambda API.
 *
 * Keep these types small and explicit: they are the only coupling between the
 * two applications and are versioned with the code.
 */

export interface HealthResponse {
  status: 'ok';
  timestamp: string;
}

/**
 * Allow-listed identity projection returned by `GET /api/me`.
 *
 * Only stable identifiers and presentation data are returned. The raw token and
 * the full claim set are never echoed back to the browser.
 */
export interface MeResponse {
  /** Stable user object identifier (`oid`). Use this for authorization. */
  userId: string;
  /** Tenant identifier (`tid`). */
  tenantId: string;
  /** Delegated scopes granted to the calling application (`scp`). */
  scopes: string[];
  /** Application roles assigned to the user (`roles`), if any. */
  roles: string[];
  /** Presentation only. Never use for authorization. */
  displayName?: string;
}

export interface ApplicationRecord {
  id: string;
  name: string;
  description?: string;
  /** `oid` of the user who owns the record. */
  ownerId: string;
  createdAt: string;
}

export interface CreateApplicationRequest {
  name: string;
  description?: string;
}

export interface ListApplicationsResponse {
  items: ApplicationRecord[];
}

export type ApiErrorCode =
  | 'VALIDATION_ERROR'
  | 'UNAUTHORIZED'
  | 'FORBIDDEN'
  | 'NOT_FOUND'
  | 'CONFLICT'
  | 'RATE_LIMITED'
  | 'INTERNAL_ERROR';

export interface ApiErrorBody {
  error: {
    code: ApiErrorCode;
    message: string;
    /** Echoed so a user can quote it to support without exposing internals. */
    correlationId: string;
    /** Field-level validation details, when applicable. */
    details?: Record<string, string>;
  };
}
